yt/data_objects/octree_subset.py

```diff
@@ -572,22 +573,67 @@ def dds(self):
     def clear_data(self):
         pass
 
-    def get_vertex_centered_data(self, *args, **kwargs):
-        raise NotImplementedError
+    def get_vertex_centered_data(self, fields, smoothed=False, no_ghost=False):
+        field = fields[0]
+        new_field = self.block_slice.get_vertex_centered_data(fields)[field]
+        return {field: new_field[..., self.ind]}
 
     @contextmanager
     def _field_parameter_state(self, field_parameters):
         yield self.block_slice.octree_subset._field_parameter_state(field_parameters)
 
 
 class OctreeSubsetBlockSlice:
-    def __init__(self, octree_subset):
+    def __init__(self, octree_subset, ds):
         self.octree_subset = octree_subset
+        self.ds = ds
+        self._vertex_centered_data = {}
         # Cache some attributes
         for attr in ["ires", "icoords", "fcoords", "fwidth"]:
             v = getattr(octree_subset, attr)
             setattr(self, "_%s" % attr, octree_subset._reshape_vals(v))
 
+    @property
+    def octree_subset_with_gz(self):
+        subset_with_gz = getattr(self, "_octree_subset_with_gz", None)
+        if not subset_with_gz:
+            self._octree_subset_with_gz = self.octree_subset.retrieve_ghost_zones(1, [])
+        return self._octree_subset_with_gz
+
+    def get_vertex_centered_data(self, fields, smoothed=False, no_ghost=False):
+        if no_ghost is True:
+            raise NotImplementedError(
+                "get_vertex_centered_data without ghost zones for oct-based datasets has not been implemented."
+            )
+
+        # Make sure the field list has only unique entries
+        fields = list(set(fields))
+        new_fields = {}
+        cg = self.octree_subset_with_gz
+        for field in fields:
+            if field in self._vertex_centered_data:
+                new_fields[field] = self._vertex_centered_data[field]
+            else:
+                finfo = self.ds._get_field_info(field)
+                orig_field = cg[field]
+                nocts = orig_field.shape[-1]
+                new_field = np.zeros((3, 3, 3, nocts), order="F")
+                new_field += orig_field[1:, 1:, 1:]
+                new_field += orig_field[:-1, 1:, 1:]
+                new_field += orig_field[1:, :-1, 1:]
+                new_field += orig_field[1:, 1:, :-1]
+                new_field += orig_field[:-1, 1:, :-1]
+                new_field += orig_field[1:, :-1, :-1]
+                new_field += orig_field[:-1, :-1, 1:]
+                new_field += orig_field[:-1, :-1, :-1]
```

Comment:
This suggestion wasn't trivial for me to write, and is maybe not that much of an improvement in terms of readability, but I do think it is slightly more worthy of confidence since it either gets everything wrong or everything right.
```suggestion
                new_field = np.zeros((3, 3, 3, nocts), order="F")
                from itertools import product, repeat
                bounds = (slice(1, None), slice(None, -1))
                for s1, s2, s3 in product(*repeat(bounds, 3)):
                    new_field += orig_field[s1, s2, s3]
```